[WIP] selection with ListSource

davidroeca opened the pull request:
This PR changes the selection implementation from a standard list to `ListSource`. It is still a WIP, in that only the dummy and gtk backends have implemented the changes at this point.

At this point, I'd like to confirm whether or not this is the ideal approach prior to building out the other backends.

Related to some discussion on https://github.com/pybee/toga/issues/539 as well as https://github.com/pybee/toga/pull/536

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] All new features have been tested
- [ ] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented on src/core/toga/widgets/selection.py:
```diff
@@ -60,7 +59,7 @@ def value(self):
     @value.setter
     def value(self, value):
-        if value not in self._items:
+        if value not in (i.field for i in self.items):
```
We should also account for the case where the user provides the actual row object (whatever the underlying item Source is storing) as the selected value. Likewise on the getter; we should be returning the underlying row object, not the display label.

davidroeca commented:
Thanks for the comments! Realized that it would also probably make sense to use the expected notify handler methods of `clear`, `insert`, `remove` to each platform implementation so that the widget is “controlled” from the source.

As for the lookups, I agree (I believe it was still a problem with the old code too but that’s beside the point).

Even though it’s called a “List” Source, what are your thoughts on using an ordered dict for storing the data (keeping the list-like interface in-tact to someone leveraging a list source)? That would resolve the linear probe issue and would maintain the same order of insertion. And maybe there’s some optional configuration of which key is stored for optimal lookups

freakboy3742 commented:
I don't have any fundamental objection to changing the underlying storage for the ListSource widget - after all, isolating the underlying data from the interface used to access that data is the entire point of the Source abstraction. However, I'm not sure how an OrderedDict would work in practice as the underlying data store - what are the key/value pairs? And how does would having lookup keyed by any value help find the *index* of the item (which is what the linear probe is currently being used to extract)?

davidroeca commented on src/core/toga/sources/list_source.py:
```diff
@@ -98,8 +98,9 @@ def append(self, *values, **named):
+        index = self._data.index(row)
         self._data.remove(row)
-        self._notify('remove', item=row)
+        self._notify('remove', index=index, item=row)
```
I'm not sure this is the best change (and I will also have to fix a few texts if we keep this). It seems that the way to remove an item in gtk is to [provide the index rather than the value](https://lazka.github.io/pgi-docs/#Gtk-3.0/classes/ComboBoxText.html#Gtk.ComboBoxText.remove).

The reason this is necessary is because the ListSource calls the `_notify` method after a removal, so there's no way to know the index of the removed item after the fact.

davidroeca commented:
@freakboy3742 I think this addresses most of the issues you've outlined above--let me know what you think.

Efficiently implementing the `remove` listener is challenging because the item is removed prior to calling the handler itself, so figuring out which item to remove on the back-end is tricky. In an outdated comment, I outlined one proposed solution where the index is passed to the remove handler. Switching to `def remove(self, index, item)` makes sense to me since this is a handler. The naming here is a bit confusing since the `remove` listener is actually closer to something like `on_remove` and the other listener APIs seem to have the same function signature. I went down this rabbit hole and realized that changing the API has broader-sweeping changes that might be worth their own PR.

davidroeca commented:
@freakboy3742 what's the preference here? There a few alternatives:

1. Retain the API and treat all mutations of the source like a full data reload (I believe this is what happens in the `detailedlist` for cocoa, winforms, etc). This isn't the most optimal, but would work for now.
2. Modify the listener API with the `index`, e.g. https://github.com/pybee/toga/pull/557
3. Keep the listener API as-is and add a lifecycle listener that lets widgets know what will happen before it does (e.g. `def will_remove(self, row)`, so that the widget can keep track of the index before the deletion actually occurs)
4. Something else?

freakboy3742 commented:
This idea is still valid, but there's been enough drift that starting from scratch is likely an easier approach. Apologies for the lack of attention that this PR received.
